torch/distributed/_shard/checkpoint/state_dict_loader.py

```diff
@@ -108,10 +114,17 @@ def load_state_dict(
     Users must call `load_state_dict` on the root module to ensure load
     pos-processing and non-tensor data properly propagates.
 
+    This function can be used for local inference and load a checkpoint
+    produced by ``save_state_dict`` without having a process group initialized
+    by passing `no_dist=True` and by using Tensors instead of ShardedTensors.
```

Comment:
I am wondering if we should have some way to avoid this `no_dist` arg, it's a bit non-intuitive to the user. How about if we do the following: We just check if there's ShardedTensor in state_dict, and if there's at least one, we require there's at least a global pg or user passed in pg, if not, we error out. If there's no ShardedTensor, we just do the behavior like `no_dist=True` is currently doing. This might avoid explicitly specify the no_dist arg.

Reply:
The problem with the `no_dist` behavior is that it changes how load/save works in a way that's not trivial to observe.

Right now, it's pretty save to assume that you can't create a `state_dict` with ShardedTensor instances and PG not initialized. That constraint is enforced by SharedTensor creation but that might be relaxed in the future.

The issue is when you're presented with a state_dict with no ShardedTensors.
For example, it's reasonable to use the API with a DDP model, for example, as it provides coordination. If we were to infer that this is a `no_dist=True` case, we'd produce multiple, possibly conflicting, checkpoints and do no synchronization across ranks.